A few bug fixes

Fesaa opened the pull request:
# Changed
- Changed: If OIDC sync is enabled, logging in will now trigger the sync (develop)
- Changed: Show K+ eligible badge for all libraries that have at least K+ metadata (Closes #3958)

# Fixed
- Fixed: Fixed the OIDC discovery document being loaded on each token refresh (develop)
- Fixed: Fixed non OIDC user being synced when OIDC sync is enabled (develop)
- Fixed: Fixed newly created reading profiles sometimes saving twice
- Fixed: Fixed the base url setting not being respected in some places (Fixes #3935)
- Fixed: Fixed page not updating in the webtoon scroller in Safari (Fixes #3754)
- Fixed: Fixed CheckUpdate getting stuck in a loop if GitHub can't be reached (Fixes #3925)
- Fixed: Fixed favicons not downloading from CoverDB (Fixes #3960)
- Fixed: Fixed the webtoon reader skipping chapters while scroller if images load slowly (Fixes #3089)
- Fixed: Fixed age rating mappings not working if any keys were equal (when normalized) (develop)

Fesaa committed:
Ensure we're not loading the discovery document for each refresh

Fesaa committed:
Do not sync non OIDC owned users

Fesaa committed:
Ensure reading profiles are not saved twice

Fesaa committed:
Fix base url missing in a lot of places Fixes #3935

Fesaa committed:
Fix webtoon scroller not updating page on webkit Fixes #3754

Fesaa committed:
Show K+ eligible for all types that support metadata Fixes #3958

Update tooltip to and/or to make it clear the library might not support
scrobbling

Fesaa committed:
Safely retry CheckUpdate calls Fixes #3925

Fesaa committed:
Fix favicon downloads not working Fixes #3960

Also trim windows newline to be safer

Fesaa committed:
Prevent next chapter from being triggered if not all images have loaded Fixes #3089

I haven't gone over the code for allImagesLoaded too much, and assume
this might cause unwanted blocks sometimes. But from reading webtoons
myself, I would prefer having to click next chapter more than scrolling
past several chapters because my wifi is being slow

Fesaa committed:
Sync user with OIDC on login

Fesaa committed:
Pick highest age rating for normalized equal mappings

Fesaa committed:
Fix unit tests

Fesaa committed:
Include barebone copilot instructions

majora2007 commented on API/Helpers/TaskHelper.cs:
```diff
@@ -0,0 +1,41 @@
+    /// <param name="action"></param>
+    /// <param name="maxAttempts"></param>
+    /// <param name="baseDelay">Base delay in ms</param>
+    public static async Task WithRetry(ILogger logger, Func<Task> action, int maxAttempts, int baseDelay)
```
Can we investigate Polly over a custom implementation?

majora2007 commented on API/Helpers/TaskHelper.cs:
```diff
@@ -0,0 +1,41 @@
+            }
+            catch (Exception e)
+            {
+                logger.LogWarning(e, "An exception occured running task {Attempt}/{MaxAttemps}", attempt+1,maxAttempts);
```
Spacing

majora2007 commented on .github/copilot-instructions.md:
```diff
@@ -0,0 +1,17 @@
+﻿## Coding Guidelines
```
Add a rule that operations (+, -, , etc) should all have spaces around them (`a + b`) not `a+b`. 

Add a rule that there should always be a space after a comma.

majora2007 commented on API/Services/OidcService.cs:
```diff
@@ -565,10 +580,10 @@ private async Task SyncAgeRestriction(OidcConfigDto settings, ClaimsPrincipal cl
     {
 
-        _discoveryDocument ??= await LoadOidcConfiguration(dto.Authority);
+        DiscoveryDocument ??= await LoadOidcConfiguration(dto.Authority);
```
This needs to be locked to avoid concurrency issues.

majora2007 commented on UI/Web/src/app/manga-reader/_components/infinite-scroller/infinite-scroller.component.ts:
```diff
@@ -220,14 +220,29 @@ export class InfiniteScrollerComponent implements OnInit, OnChanges, OnDestroy,
+      fromEvent(element, 'scrollend')
+        .pipe(debounceTime(20), takeUntilDestroyed(this.destroyRef))
+        .subscribe((event) => {
+          this.handleScrollEndEvent(event)
```
Move this into a tap and keep subscribe empty. This is the "best approach" i learned after this code was written.

Fesaa replied:
I'm curious what the reason behind this is

majora2007 replied:
I think because everything is within the pipe.

majora2007 commented on UI/Web/src/app/manga-reader/_components/infinite-scroller/infinite-scroller.component.ts:
```diff
@@ -220,14 +220,29 @@ export class InfiniteScrollerComponent implements OnInit, OnChanges, OnDestroy,
+    const isScrollEndSupported = 'onscrollend' in document;
+    if (isScrollEndSupported) {
+      fromEvent(element, 'scrollend')
+        .pipe(debounceTime(20), takeUntilDestroyed(this.destroyRef))
```
Don't hardcode magic numbers. Put this in a const above the class with a descriptive name.

majora2007 commented on UI/Web/src/app/manga-reader/_components/infinite-scroller/infinite-scroller.component.ts:
```diff
@@ -220,14 +220,29 @@ export class InfiniteScrollerComponent implements OnInit, OnChanges, OnDestroy,
+      });
+
+    const isScrollEndSupported = 'onscrollend' in document;
+    if (isScrollEndSupported) {
```
Don't duplicate the logic, just change the string we build the fromEvent from.

majora2007 commented on UI/Web/src/app/user-settings/manage-reading-profiles/manage-reading-profiles.component.ts:
```diff
@@ -219,41 +221,50 @@ export class ManageReadingProfilesComponent implements OnInit {
   }
 
+  /**
+   * Autosave will not error
```
What does this comment mean if we are using catchError?

Fesaa replied:
That the method itself will not throw errors / handle them safely. Means it's safe to set the signal back to false above

majora2007 replied:
I would elaborate more. It's not clear the intent of the comment and leads to confusion.

majora2007 commented on UI/Web/src/app/user-settings/user-holds/scrobbling-holds.component.html:
```diff
@@ -17,7 +17,7 @@
       <ng-template let-item="row" ngx-datatable-cell-template>
         <app-image [width]="'32px'" [height]="'32px'" [imageUrl]="imageService.getSeriesCoverImage(item.seriesId)"></app-image>
-        <a class="btn-link ms-2" href="/library/{{item.libraryId}}/series/{{item.seriesId}}" target="_blank">{{item.seriesName}}</a>
+        <a class="btn-link ms-2" href="{{baseUrl}}library/{{item.libraryId}}/series/{{item.seriesId}}" target="_blank">{{item.seriesName}}</a>
```
We should add a copilot rule to ensure any hrefs start with baseUrl

Fesaa committed:
Copilot typo's and PR comments

Fesaa committed:
Use polly instead of own retry method

Fesaa committed:
Include tooltip explaining identity provider icons

Fesaa committed:
Fix preference side nav not being shown when being redirected to settings on login

Race condition where the obserable was never subscribed to before the
route, so would never get the event.
And then angular reactivity not updating. Loading them with let
beforehand forces both

majora2007 commented on API/Services/OidcService.cs:
```diff
@@ -631,18 +634,37 @@ private static async Task<ClaimsPrincipal> ParseIdToken(OidcConfigDto dto, strin
+                return _discoveryDocument;
+            }
+
+            var hasTrailingSlash = authority.EndsWith('/');
```
I don't see you ever setting or accessing _discoveryDocument within the lock

majora2007 commented on UI/Web/src/app/admin/manage-users/manage-users.component.html:
```diff
@@ -25,10 +25,10 @@
               @switch (member.identityProvider) {
                 @case (IdentityProvider.OpenIdConnect) {
-                  <app-image imageUrl="assets/icons/open-id-connect-logo.svg" height="16px" width="16px"></app-image>
+                  <app-image [ngbTooltip]="t('oidc')" imageUrl="assets/icons/open-id-connect-logo.svg" height="16px" width="16px"></app-image>
```
key should be identity-provider-oidc-tooltip

majora2007 commented on UI/Web/src/app/admin/manage-users/manage-users.component.html:
```diff
@@ -25,10 +25,10 @@
                 }
                 @case (IdentityProvider.Kavita) {
-                  <app-image imageUrl="assets/icons/favicon-16x16.png" height="16px" width="16px"></app-image>
+                  <app-image [ngbTooltip]="t('native')" imageUrl="assets/icons/favicon-16x16.png" height="16px" width="16px"></app-image>
```
key should be identity-provider-native-tooltip

majora2007 commented on UI/Web/src/app/manga-reader/_components/infinite-scroller/infinite-scroller.component.ts:
```diff
@@ -179,6 +179,15 @@ export class InfiniteScrollerComponent implements OnInit, OnChanges, OnDestroy,
+  /**
+   * Default debounce time from scroll and scrollend event listeners
+   */
+  defaultScrollDebounce = 20;
```
Make these const, which means outside the class. They go above the class declaration.

majora2007 commented on UI/Web/src/app/manga-reader/_components/infinite-scroller/infinite-scroller.component.ts:
```diff
@@ -223,26 +232,24 @@ export class InfiniteScrollerComponent implements OnInit, OnChanges, OnDestroy,
-      .subscribe((event) => {
-        this.handleScrollEvent(event)
-      });
+      .pipe(
```
The original code has scrollend calling handleScrollEndEvent(), your new code has them both calling handleScrollEvent

Fesaa committed:
Some more comments

Fesaa commented on UI/Web/src/app/user-settings/manage-reading-profiles/manage-reading-profiles.component.ts:
```diff
@@ -219,41 +221,46 @@ export class ManageReadingProfilesComponent implements OnInit {
-          this.toastr.error(err.message);
-        }
-      })
+          return of(null);
```
The return value isn't actually used, it just needs to be an observable to fit in the pipe

Fesaa commented on UI/Web/src/app/user-settings/manage-reading-profiles/manage-reading-profiles.component.ts:
```diff
@@ -219,41 +221,46 @@ export class ManageReadingProfilesComponent implements OnInit {
+      catchError(err => {
+        console.log(err);
+        this.toastr.error(err.message);
+        return of(null);
```
The return value isn't actually used, it just needs to be an observable to fit in the pipe

majora2007 replied:
I guess this can't detect the lack of newline above the return. You have multiple cases of this in the PR.

Fesaa replied:
Maybe it only works once the document in the main branch? I'll find the missing newlines

Fesaa commented on API/Services/Plus/ExternalMetadataService.cs:
```diff
@@ -1806,10 +1806,15 @@ public static AgeRating DetermineAgeRating(IEnumerable<string> values, Dictionar
+            );
 
         return values
-            .Select(v => mappings.TryGetValue(v.ToNormalized(), out var mapping) ? mapping : AgeRating.Unknown)
+            .Select(v => mappings.GetValueOrDefault(v.ToNormalized(), AgeRating.Unknown))
```
Yes, each value is normalized once, as well as each key in the dictionary. This is correct

Fesaa commented on API/Services/OidcService.cs:
```diff
@@ -68,7 +69,9 @@ public class OidcService(ILogger<OidcService> logger, UserManager<AppUser> userM
 
-    private OpenIdConnectConfiguration? _discoveryDocument;
+    private static OpenIdConnectConfiguration? _discoveryDocument;
+    private static readonly SemaphoreSlim _discoveryDocumentSemaphore = new(1, 1);
```
There will only ever be one authority, this won't be a problem. We're also not unit testing anything with the document directly.

Fesaa committed:
Newlines, ConfigurationManager, and clearify what the Authority URL should be
